### PR TITLE
Implement str.isalpha() based on CPython

### PR DESCRIPTION
**Pull request by 1e-to**

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Thanks for the PR, I've queued it for review.

File: numba/tests/test_unicode.py [hunk 1]
```diff
@@ -1277,6 +1277,24 @@ def pyfunc(x):
...
+        def pyfunc(x):
+            return x.isalpha()
+
+        cfunc = njit(pyfunc)
+        # Samples taken from CPython testing:
+        # https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Lib/test/test_unicode.py#L738-745     # noqa: E501
+        cpython = ['\uD800', '\uDFFF', '\uD800\uD800', '\uDFFF\uDFFF',
+                   'a\uD800b\uDFFF', 'a\uDFFFb\uD800',
+                   'a\uD800b\uDFFFa', 'a\uDFFFb\uD800a']
+        # https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Lib/test/test_unicode.py#L613-L621    # noqa: E501
```
Comment (stuartarchibald) [inline numba/tests/test_unicode.py]: This URL refers to `test_isspace`, this code is checking `isalpha`

File: numba/tests/test_unicode.py [hunk 2]
```diff
@@ -1277,6 +1277,24 @@ def pyfunc(x):
...
+        extras = ['\u2167', '\u2177', '\U00010401', '\U00010427',
```
Comment (stuartarchibald) [inline numba/tests/test_unicode.py]: These extra values correspond to testing `isupper/islower` not `isalpha`

File: numba/unicode.py [hunk 3]
```diff
@@ -1267,6 +1268,26 @@ def impl(a):
...
+# https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L11928-L11964    # noqa: E501
+@overload_method(types.UnicodeType, 'isalpha')
+def unicode_isalnum(data):
+    """Implements UnicodeType.isalpha()"""
+
+    def impl(data):
+        length = len(data)
+        if length == 0:
+            return False
+
```
Comment (stuartarchibald) [inline numba/unicode.py]: This probably ought to have an ASCII only fast-path? Hint `Py_ISALPHA`.

Comment (stuartarchibald) [reply numba/unicode.py]: This is also missing the single character case shortcut

File: numba/tests/test_unicode.py (hunk 1 above)
Comment (1e-to) [reply numba/tests/test_unicode.py]: Fixed

File: numba/tests/test_unicode.py (hunk 2 above)
Comment (1e-to) [reply numba/tests/test_unicode.py]: Fixed

File: numba/unicode.py (hunk 3 above)
Comment (1e-to) [reply numba/unicode.py]: the single character case shortcut not nessesary, i implement for 1 and more characters case

Comment (stuartarchibald) [reply numba/unicode.py]: > the single character case shortcut not nessesary, i implement for 1 and more characters case

I feel we need to understand the reason it exits in CPython before deciding on it's necessity. Do you have any ideas?

File: numba/tests/test_unicode.py [hunk 4]
```diff
@@ -1277,6 +1277,24 @@ def pyfunc(x):
             self.assertEqual(pyfunc(*args), cfunc(*args),
                              msg='failed on {}'.format(args))
 
+    def test_isalpha(self):
+        def pyfunc(x):
+            return x.isalpha()
+
+        cfunc = njit(pyfunc)
+        # Samples taken from CPython testing:
+        # https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Lib/test/test_unicode.py#L630-640    # noqa: E501
```
Comment (stuartarchibald) [inline numba/tests/test_unicode.py]:
```suggestion
        # https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Lib/test/test_unicode.py#L630-L640    # noqa: E501
```
This syntax is invalid, `L` prefix needed ahead of both items in the range.

File: numba/tests/test_unicode.py [hunk 5]
```diff
@@ -1277,6 +1277,24 @@ def pyfunc(x):
...
+        cpython = ['\u1FFc', '\U00010401', '\U00010427', '\U00010429',
+                   '\U0001044E', '\U0001F40D', '\U0001F46F']
+        # https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Lib/test/test_unicode.py#L738-745    # noqa: E501
```
Comment (stuartarchibald) [inline numba/tests/test_unicode.py]:
```suggestion
        # https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Lib/test/test_unicode.py#L738-L745    # noqa: E501
```

File: numba/unicode.py (hunk 3 above)
Comment (1e-to) [reply numba/unicode.py]: This is necessary for a short path, as well as with an empty string or with a check for asсii
I was wrong. I added a one-character string check to the implementation

**Comment by 1e-to** [comment]
Add to queue for review, please

**Comment by stuartarchibald** [comment]
Thanks for the fixes, marking for merge.

**Comment by stuartarchibald** [comment]
Please could you resolve the conflicts by merging master into this branch. Thanks!

**Comment by 1e-to** [comment]
> Please could you resolve the conflicts by merging master into this branch. Thanks!

Done.

**Pull request merged**